File: platform/jvm/capture/src/main/kotlin/io/bitdrift/capture/utils/ConfigCache.kt
// capture-sdk - bitdrift's client SDK
// Copyright Bitdrift, Inc. All rights reserved.
//
// Use of this source code is governed by a source available license that can be found in the
// LICENSE file or at:
// https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt

package io.bitdrift.capture.utils

import androidx.annotation.VisibleForTesting
import java.io.File

internal class CacheFormattingError : Exception()

/**
 * Parses per-line, comma-delimited key/value pairs into configuration values
 */
internal object ConfigCache {
    /**
     * Read configuration values from a file
     *
     * @param file File to read
     *
     * @return The values or Exception which occurred during parsing
     *
     * @throws java.io.IOException, CacheFormattingError
     */
    fun readValues(file: File): Map<String, Any> = readValues(file.readText())

    /**
     * Read configuration values from text
     *
     * @param text to parse
     *
     * @return The values or Exception which occurred during parsing
     *
     * @throws CacheFormattingError if the file format does not match
     */
    @VisibleForTesting(otherwise = VisibleForTesting.PRIVATE)
    fun readValues(text: String): Map<String, Any> =
        text
            .lines()
            .map { line ->
                val pair = line.split(",", limit = 2)
                if (pair.size == 2) {
                    Pair(
                        pair[0],
                        when (pair[1]) {
                            "true" -> true
                            "false" -> false
                            else -> pair[1]
                        },
                    )
                } else {
                    throw CacheFormattingError()
                }
            }.associate { it.first to it.second }
}

File: platform/swift/source/reports/ConfigCache.swift
// capture-sdk - bitdrift's client SDK
// Copyright Bitdrift, Inc. All rights reserved.
//
// Use of this source code is governed by a source available license that can be found in the
// LICENSE file or at:
// https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt

/// Parse text for comma-delimited key/value pairs
///
/// - parameter text: text to parse
///
/// - returns: A mapping between keys and values, upon success
package func readCachedValues(_ text: String) -> [String: Any]? {
    var values: [String: Any] = [:]
    for line in text.split(separator: "\n") {
        let pair = line.split(separator: ",", maxSplits: 1)
        if pair.count == 2 {
            let key = String(pair[0])
            switch pair[1] {
            case "true":
                values[key] = true
            case "false":
                values[key] = false
            default:
                values[key] = String(pair[1])
            }
        } else {
            // failed to fully parse file, contains errors
            return nil
        }
    }
    return values
}
